jcl/src/java.base/share/classes/java/lang/JFRHelpers.java

```diff
@@ -42,51 +42,101 @@ final class JFRHelpers {
 	private static volatile boolean jfrClassesInitialized = false;
 	private static String jfrCMDLineOption = com.ibm.oti.vm.VM.getjfrCMDLineOption();
 
-	private static Long convertToBytes(String sizeValue) {
-		long sizeInBytes = 0L;
-		String numericPart = sizeValue.replaceAll("[^0-9]", "");
-		String sizeUnit = sizeValue.replaceAll("[0-9]", "").toLowerCase();
-		long size = Long.parseLong(numericPart);
-		switch (sizeUnit) {
-		case "k": /* intentional fall through - KiloBytes */
-			sizeInBytes = size * 1_024L;
+	private static int indexOfUnit(String text) {
+		for (int i = 0; i < text.length(); i++) {
+			char c = text.charAt(i);
+			if (i == 0 && c == '-') { // Accept negative values.
+				continue;
+			}
+			if (!Character.isDigit(c)) {
+				return i;
+			}
+		}
+		return text.length();
+	}
+
+	private static Long convertToBytes(String text) {
+		if (null == text) {
+			throw new IllegalArgumentException("Parsing error memory size value: syntax error, value is null");
+		}
+		int index = indexOfUnit(text);
+		String textValue = text.substring(0, index);
+		String unit = text.substring(index);
+		long bytes = 0L;
+		try {
+			bytes = Long.parseLong(textValue);
+		} catch (NumberFormatException nfe) {
+			throw new IllegalArgumentException("Parsing error memory size value: invalid value");
+		}
+		if (bytes < 0) {
+			throw new IllegalArgumentException("Parsing error memory size value: negative values not allowed");
+		}
+		if (unit.isEmpty()) {
+			return bytes;
+		}
```

Comment:
This can be greatly simplified by using `Pattern`:
```java
		Pattern pattern = Pattern.compile("(\\d+)([gkm]b?)?", Pattern.CASE_INSENSITIVE);
		Matcher matcher = pattern.matcher(text);
		if (!matcher.matches()) {
			throw new IllegalArgumentException("Invalid memory size");
		}
		String textValue = matcher.group(1);
		long bytes;
		try {
			bytes = Long.parseLong(textValue);
		} catch (NumberFormatException nfe) {
			throw new IllegalArgumentException("Invalid memory size");
		}
		String unit = matcher.group(2);
		if ((unit != null) && !unit.isEmpty()) {
			switch (unit.toLowerCase()) {
			case "k":
			case "kb":
				bytes *= 1024L;
				break;
...
```